File: src-tauri/src/application_menu.rs
use tauri::{
    menu::{CheckMenuItem, Menu, MenuBuilder, MenuItem, Submenu, SubmenuBuilder},
    AppHandle, Result, Wry,
};

use crate::window::create_window;

fn build_app_menu(app_handle: &AppHandle) -> Result<Submenu<Wry>> {
    let settings = MenuItem::with_id(app_handle, "settings", "Settings...", true, Some("Cmd+,"))?;

    SubmenuBuilder::new(app_handle, "App")
        .about(None)
        .separator()
        .item(&settings)
        .separator()
        .services()
        .separator()
        .hide()
        .hide_others()
        .show_all()
        .separator()
        .quit()
        .separator()
        .build()
}

fn build_file_menu(app_handle: &AppHandle) -> Result<Submenu<Wry>> {
    let new_window =
        MenuItem::with_id(app_handle, "new_window", "New Window", true, Some("Cmd+N"))?;
    let new_tab = MenuItem::with_id(app_handle, "new_tab", "New Tab", true, Some("Cmd+T"))?;
    let close_window = MenuItem::with_id(
        app_handle,
        "close_window",
        "Close Window",
        true,
        Some("Cmd+Shift+W"),
    )?;
    let close_tab = MenuItem::with_id(app_handle, "close_tab", "Close Tab", true, Some("Cmd+W"))?;

    let open = MenuItem::with_id(app_handle, "open", "Open", true, Some("Cmd+O"))?;
    let rename = MenuItem::with_id(app_handle, "rename", "Rename", true, None::<&str>)?;
    let move_to_trash = MenuItem::with_id(
        app_handle,
        "move_to_trash",
        "Move to Trash",
        true,
        Some("Cmd+Backspace"),
    )?;

    SubmenuBuilder::new(app_handle, "File")
        .item(&new_window)
        .item(&new_tab)
        .item(&close_window)
        .item(&close_tab)
        .separator()
        .item(&open)
        .item(&rename)
        .item(&move_to_trash)
        .build()
}

fn build_edit_menu(app_handle: &AppHandle) -> Result<Submenu<Wry>> {
    let cut = MenuItem::with_id(app_handle, "cut", "Cut", true, Some("Cmd+X"))?;
    let copy = MenuItem::with_id(app_handle, "copy", "Copy", true, Some("Cmd+C"))?;
    let paste = MenuItem::with_id(app_handle, "paste", "Paste", true, Some("Cmd+V"))?;
    let select_all =
        MenuItem::with_id(app_handle, "select_all", "Select All", true, Some("Cmd+A"))?;
    let find = MenuItem::with_id(app_handle, "find", "Find", true, Some("Cmd+F"))?;

    SubmenuBuilder::new(app_handle, "Edit")
        .undo()
        .redo()
        .separator()
        .item(&cut)
        .item(&copy)
        .item(&paste)
        .item(&select_all)
        .separator()
        .item(&find)
        .build()
}

fn build_view_menu(app_handle: &AppHandle) -> Result<Submenu<Wry>> {
    let view_as_list = CheckMenuItem::with_id(
        app_handle,
        "view_as_list",
        "as List",
        true,
        false,
        Some("Cmd+1"),
    )?;
    let view_as_thumbnail = CheckMenuItem::with_id(
        app_handle,
        "view_as_thumbnail",
        "as Thumbnail",
        true,
        false,
        Some("Cmd+2"),
    )?;
    let view_as_gallery = CheckMenuItem::with_id(
        app_handle,
        "view_as_gallery",
        "as Gallery",
        true,
        false,
        Some("Cmd+3"),
    )?;

    let toggle_navigator = MenuItem::with_id(
        app_handle,
        "toggle_navigator",
        "Toggle Navigator",
        true,
        None::<&str>,
    )?;
    let toggle_inspector = MenuItem::with_id(
        app_handle,
        "toggle_inspector",
        "Toggle Inspector",
        true,
        None::<&str>,
    )?;

    let refresh = MenuItem::with_id(app_handle, "refresh", "Refresh", true, Some("Cmd+R"))?;
    let force_reload = MenuItem::with_id(
        app_handle,
        "force_reload",
        "Force Reload",
        true,
        Some("Cmd+Shift+R"),
    )?;
    let toggle_developer_tools = MenuItem::with_id(
        app_handle,
        "toggle_developer_tools",
        "Toggle Developer Tools",
        true,
        Some("Cmd+Option+I"),
    )?;

    let sort_by_name = CheckMenuItem::with_id(
        app_handle,
        "sort_by_name",
        "Name",
        true,
        false,
        None::<&str>,
    )?;
    let sort_by_date_created = CheckMenuItem::with_id(
        app_handle,
        "sort_by_date_created",
        "Date Created",
        true,
        false,
        None::<&str>,
    )?;
    let sort_by_date_modified = CheckMenuItem::with_id(
        app_handle,
        "sort_by_date_modified",
        "Date Modified",
        true,
        false,
        None::<&str>,
    )?;
    let sort_by_date_last_opened = CheckMenuItem::with_id(
        app_handle,
        "sort_by_date_last_opened",
        "Date Opened",
        true,
        false,
        None::<&str>,
    )?;
    let sort_by_size = CheckMenuItem::with_id(
        app_handle,
        "sort_by_size",
        "Size",
        true,
        false,
        None::<&str>,
    )?;
    let sort_by_rating = CheckMenuItem::with_id(
        app_handle,
        "sort_by_rating",
        "Rating",
        true,
        false,
        None::<&str>,
    )?;
    let sort_menu = SubmenuBuilder::new(app_handle, "Sort By")
        .items(&[
            &sort_by_name,
            &sort_by_date_created,
            &sort_by_date_modified,
            &sort_by_date_last_opened,
            &sort_by_size,
            &sort_by_rating,
        ])
        .build()?;

    SubmenuBuilder::new(app_handle, "View")
        .item(&view_as_list)
        .item(&view_as_thumbnail)
        .item(&view_as_gallery)
        .separator()
        .item(&sort_menu)
        .separator()
        .item(&toggle_navigator)
        .item(&toggle_inspector)
        .separator()
        .item(&refresh)
        .separator()
        // TODO: force reload, devtools
        .item(&force_reload)
        .item(&toggle_developer_tools)
        .separator()
        .fullscreen()
        .build()
}

fn build_go_menu(app_handle: &AppHandle) -> Result<Submenu<Wry>> {
    let back = MenuItem::with_id(app_handle, "back", "Back", true, Some("Cmd+["))?;
    let forward = MenuItem::with_id(app_handle, "forward", "Forward", true, Some("Cmd+]"))?;

    SubmenuBuilder::new(app_handle, "Go")
        .item(&back)
        .item(&forward)
        .build()
}

fn build_window_menu(app_handle: &AppHandle) -> Result<Submenu<Wry>> {
    SubmenuBuilder::new(app_handle, "Window")
        .minimize()
        // TODO: zoom, front, window
        .build()
}

pub fn build_menu(app_handle: &AppHandle) -> Result<Menu<Wry>> {
    let app_menu = build_app_menu(app_handle)?;
    let file_menu = build_file_menu(app_handle)?;
    let edit_menu = build_edit_menu(app_handle)?;
    let view_menu = build_view_menu(app_handle)?;
    let go_menu = build_go_menu(app_handle)?;
    let window_menu = build_window_menu(app_handle)?;
    // TODO: help

    MenuBuilder::new(app_handle)
        .items(&[
            &app_menu,
            &file_menu,
            &edit_menu,
            &view_menu,
            &go_menu,
            &window_menu,
        ])
        .build()
}

pub fn setup_menu_events(app_handle: &AppHandle) {
    app_handle.on_menu_event(move |app_handle, event| {
        println!("menu event: {:?}", event.id());

        match event.id().as_ref() {
            "new_window" => {
                let _ = create_window(app_handle);
            }
            _ => {
                println!("unexpected menu event");
            }
        }
    });
}

File: src-electron/application-menu.ts
import {
  app,
  BrowserWindow,
  type IpcMainEvent,
  ipcMain,
  Menu,
  type MenuItemConstructorOptions,
  shell,
} from 'electron'

type State = {
  canBack: boolean
  canCloseTab: boolean
  canForward: boolean
  inspectorHidden: boolean
  navigatorHidden: boolean
  orderBy:
    | 'dateCreated'
    | 'dateLastOpened'
    | 'dateModified'
    | 'name'
    | 'score'
    | 'size'
  viewMode: 'gallery' | 'list' | 'thumbnail'
}

export type ApplicationMenuParams = Partial<State>

const isEditType = (
  type: string,
): type is 'copy' | 'cut' | 'paste' | 'selectAll' =>
  ['copy', 'cut', 'paste', 'selectAll'].includes(type)

// biome-ignore lint/suspicious/noExplicitAny: false positive
const send = (message: { type: string; data?: any }) => {
  const activeWindow = BrowserWindow.getFocusedWindow()
  if (!activeWindow) {
    return
  }

  const type = message.type
  if (isEditType(type) && activeWindow.webContents.isDevToolsFocused()) {
    activeWindow.webContents.devToolsWebContents?.[type]()
    return
  }

  activeWindow.webContents.send('onMessage', message)
}

const registerApplicationMenu = (
  createWindow: (directoryPath?: string) => void,
) => {
  const isMac = process.platform === 'darwin'

  let state: State = {
    canBack: false,
    canCloseTab: false,
    canForward: false,
    inspectorHidden: false,
    navigatorHidden: false,
    orderBy: 'name',
    viewMode: 'list',
  }

  const update = () => {
    // @see https://www.electronjs.org/docs/latest/api/menu#examples
    const template: MenuItemConstructorOptions[] = [
      // { role: 'appMenu' }
      ...(isMac
        ? [
            {
              label: app.name,
              submenu: [
                { role: 'about' },
                { type: 'separator' },
                {
                  accelerator: 'CmdOrCtrl+,',
                  click: () => send({ type: 'goToSettings' }),
                  label: 'Preferences...',
                },
                { type: 'separator' },
                { role: 'services' },
                { type: 'separator' },
                { role: 'hide' },
                { role: 'hideOthers' },
                { role: 'unhide' },
                { type: 'separator' },
                { role: 'quit' },
              ],
            } as MenuItemConstructorOptions,
          ]
        : []),
      // { role: 'fileMenu' }
      {
        label: 'File',
        submenu: [
          {
            accelerator: 'CmdOrCtrl+N',
            click: () => createWindow(),
            label: 'New Window',
          },
          {
            accelerator: 'CmdOrCtrl+t',
            click: () =>
              send({ type: 'newTab', data: { path: app.getPath('home') } }),
            label: 'New Tab',
          },
          ...[
            isMac
              ? {
                  accelerator: 'CmdOrCtrl+Shift+w',
                  role: 'close',
                }
              : { role: 'quit' },
          ],
          {
            accelerator: 'CmdOrCtrl+w',
            click: () => send({ type: 'closeTab' }),
            enabled: state.canCloseTab,
            label: 'Close Tab',
          },
          { type: 'separator' },
          {
            accelerator: 'CmdOrCtrl+O',
            click: () => send({ type: 'open' }),
            label: 'Open',
          },
          {
            click: () => send({ type: 'rename' }),
            label: 'Rename',
          },
          {
            accelerator: 'CmdOrCtrl+Backspace',
            click: () => send({ type: 'moveToTrash' }),
            label: 'Move to Trash',
          },
        ],
      } as MenuItemConstructorOptions,
      // { role: 'editMenu' }
      {
        label: 'Edit',
        submenu: [
          { role: 'undo' },
          { role: 'redo' },
          { type: 'separator' },
          {
            accelerator: 'CmdOrCtrl+X',
            click: () => send({ type: 'cut' }),
            label: 'Cut',
          },
          {
            accelerator: 'CmdOrCtrl+C',
            click: () => send({ type: 'copy' }),
            label: 'Copy',
          },
          {
            accelerator: 'CmdOrCtrl+V',
            click: () => send({ type: 'paste' }),
            label: 'Paste',
          },
          {
            accelerator: 'CmdOrCtrl+A',
            click: () => send({ type: 'selectAll' }),
            label: 'Select All',
          },
          { type: 'separator' },
          {
            accelerator: 'CmdOrCtrl+F',
            click: () => send({ type: 'find' }),
            label: 'Find',
          },
        ],
      } as MenuItemConstructorOptions,
      // { role: 'viewMenu' }
      {
        label: 'View',
        submenu: [
          ...[
            {
              accelerator: 'CmdOrCtrl+1',
              label: 'as List',
              viewMode: 'list',
            },
            {
              accelerator: 'CmdOrCtrl+2',
              label: 'as Thumbnail',
              viewMode: 'thumbnail',
            },
            {
              accelerator: 'CmdOrCtrl+3',
              label: 'as Gallery',
              viewMode: 'gallery',
            },
          ].map((menu) => ({
            ...menu,
            checked: menu.viewMode === state.viewMode,
            click: () =>
              send({
                type: 'changeViewMode',
                data: { viewMode: menu.viewMode },
              }),
            type: 'radio',
          })),
          { type: 'separator' },
          {
            label: 'Sort By',
            submenu: [
              { label: 'Name', orderBy: 'name' },
              { label: 'Date Created', orderBy: 'dateCreated' },
              { label: 'Date Modified', orderBy: 'dateModified' },
              { label: 'Date Last Opened', orderBy: 'dateLastOpened' },
              { label: 'Size', orderBy: 'size' },
              { label: 'Rating', orderBy: 'score' },
            ].map((menu) => ({
              ...menu,
              checked: menu.orderBy === state.orderBy,
              click: () =>
                send({ type: 'sort', data: { orderBy: menu.orderBy } }),
              type: 'radio',
            })),
          },
          { type: 'separator' },
          {
            label: state.navigatorHidden ? 'Show Navigator' : 'Hide Navigator',
            click: () =>
              send({
                type: 'changeSidebarHidden',
                data: {
                  variant: 'primary',
                  hidden: !state.navigatorHidden,
                },
              }),
          },
          {
            label: state.inspectorHidden ? 'Show Inspector' : 'Hide Inspector',
            click: () =>
              send({
                type: 'changeSidebarHidden',
                data: {
                  variant: 'secondary',
                  hidden: !state.inspectorHidden,
                },
              }),
          },
          { type: 'separator' },
          {
            accelerator: 'CmdOrCtrl+R',
            label: 'Refresh',
            click: () => send({ type: 'refresh' }),
          },
          { type: 'separator' },
          // { role: 'reload' },
          { role: 'forceReload' },
          { role: 'toggleDevTools' },
          { type: 'separator' },
          { role: 'togglefullscreen' },
        ],
      } as MenuItemConstructorOptions,
      {
        label: 'Go',
        submenu: [
          {
            accelerator: 'CmdOrCtrl+[',
            click: () => send({ type: 'back' }),
            enabled: state.canBack,
            label: 'Back',
          },
          {
            accelerator: 'CmdOrCtrl+]',
            click: () => send({ type: 'forward' }),
            enabled: state.canForward,
            label: 'Forward',
          },
        ],
      } as MenuItemConstructorOptions,
      // { role: 'windowMenu' }
      {
        label: 'Window',
        submenu: [
          { role: 'minimize' },
          { role: 'zoom' },
          ...(isMac
            ? [
                { type: 'separator' },
                { role: 'front' },
                { type: 'separator' },
                { role: 'window' },
              ]
            : [{ role: 'close' }]),
        ],
      } as MenuItemConstructorOptions,
      {
        role: 'help',
        submenu: [
          {
            label: 'Learn More',
            click: () => shell.openExternal('https://github.com/fiahfy/zephy'),
          },
        ],
      },
    ]

    const menu = Menu.buildFromTemplate(template)
    Menu.setApplicationMenu(menu)
  }

  update()

  ipcMain.on(
    'update',
    (_event: IpcMainEvent, params: ApplicationMenuParams) => {
      state = { ...state, ...params }
      update()
    },
  )
}

export default registerApplicationMenu
